File: algorithms/2_medium/N/01498_number_of_subsequences_that_satisfy_the_given_sum_condition/solution.c
#define MOD 1000000007

inline static int64_t pow_with_mod(int64_t base, int64_t pow, const int64_t mod_by) {
    int64_t res = 1;
    while (pow > 0) {
        if ((pow & 1) == 1) {
            res = (res * base) % mod_by;
        }
        base = (base * base) % mod_by;
        pow >>= 1;
    }
    return res;
}

static int cmp(const void *a_ptr, const void *b_ptr) {
    const int a = *(int*)a_ptr;
    const int b = *(int*)b_ptr;

    return (a > b) - (a < b);
}

int numSubseq(const int *restrict nums, const int len, const int target) {
    if (nums == NULL || len <= 0) { return 0; }

    qsort(nums, (size_t)len, sizeof(*nums), cmp);
    int64_t cnt = 0;
    int64_t lo = 0;
    int64_t hi = len - 1;

    while (lo <= hi) {
        if (nums[lo] + nums[hi] > target) {
            hi -= 1;
        } else {
            cnt = (cnt + pow_with_mod(2, hi - lo, MOD)) % MOD;
            lo += 1;
        }
    }

    return (int)cnt;
}

File: algorithms/2_medium/N/01498_number_of_subsequences_that_satisfy_the_given_sum_condition/solution.rs
impl Solution {
    const MOD: isize = 1_000_000_007;

    #[inline(always)]
    fn pow_with_mod<const MOD: isize>(mut base: isize, mut pow: isize) -> isize {
        let mut res = 1_isize;
        while pow > 0 {
            if (pow & 1) == 1 {
                res = (res * base) % MOD;
            }
            base = (base * base) % MOD;
            pow >>= 1;
        }
        res
    }

    pub fn num_subseq(mut nums: Vec<i32>, target: i32) -> i32 {
        if nums.is_empty() { return 0; }

        nums.sort_unstable();
        let len = nums.len();
        let mut cnt = 0_isize;
        let mut lo = 0_isize;
        let mut hi = len as isize - 1;

        while lo <= hi {
            if nums[lo as usize] + nums[hi as usize] > target {
                hi -= 1;
            } else {
                cnt = (cnt + Self::pow_with_mod::<{ Self::MOD }>(2, hi - lo)) % Self::MOD;
                lo += 1;
            }
        }

        cnt as i32
    }
}
